### DISCUSSION TITLE
Switching from TeeChart to ScottPlot, BlueM.Wave v3

### DISCUSSION THREAD

**Discussion by jamaa** [bluemodel/BlueM.Wave#144] [Announcements]
The backbone of Wave's charting abilities has always been Steema TeeChart.

Switching to an open source plotting library has been on the todo list for a long time, mostly because of the license fees (#70). Additionally, recent experience has shown that TeeChart development is unfortunately relatively unreliable (#122).

For these reasons, I have decided to start work on switching to a new charting library, the very impressive and open-source [Scottplot](https://github.com/ScottPlot/ScottPlot) library.

The main advantage of TeeChart over other solutions has always been the built-in editor, which will unfortunately have to be replaced using our own UI. It will probably not be feasible to replace all of the features that the TeeChart editor provided (at least not immediately), but the goal is to have at least the most important chart editing features in place, which in my view are axis settings and series styling options.

The switch to Scottplot will be released as a new major version 3 of BlueM.Wave and will probably also involve some significant changes to the UI (see screenshot below for the current state). I am planning to hopefully release some beta/preview releases soon. The development is currently taking place in #134.

As always, feedback and contributions of any kind are always appreciated!

![grafik](https://github.com/bluemodel/BlueM.Wave/assets/90166/50a546f5-a92e-45fa-99a9-61a9f7f782a4)

**Comment by wassermax** [comment]
That's indeed quite a big change! My question is, why don't we switch to Python and a slightly bigger (aka better future-proof) library?
I would have spontaneously thought of Plotly, for example:
[https://plotly.com/python/time-series/](https://plotly.com/python/time-series/)
[https://towardsdatascience.com/introduction-to-interactive-time-series-visualizations-with-plotly-in-python-d3219eb7a7af](https://towardsdatascience.com/introduction-to-interactive-time-series-visualizations-with-plotly-in-python-d3219eb7a7af)
But apparently the decision in favor of Scottplott has already been made, which is ok from my side :-)

**Reply by jamaa** [reply]
I'm a big fan of Python and use it extensively, also for plotting time series (mostly with matplotlib), but I think it is more useful when you are (interactively) coding in Python already anyway, and need some quick graphs. If you are using Python, you already have access to amazing libraries such as pandas and co, which easily cover most of the analysis functions that Wave offers and more. There are even libraries for handling hydrological metadata (check out the promising [HydroPandas](https://github.com/ArtesiaWater/hydropandas)). 

Basically, the only thing "missing" from Wave in the python ecosystem are methods for reading/writing special time series formats such as ZRXP, UVF, etc. And the user interface of course.

To me, Wave fulfills a different purpose. It is a desktop app usable with minimal setup and without requiring any programming knowledge. The analogy is kind of like using QGIS for GIS data instead of gdal on the command-line.

Of course, it's also possible to create a fully-fledged desktop app including a user interface using Python, but it would basically mean having to rewrite everything, starting from scratch. Also, there is no easy choice for the GUI frontend (PyQt maybe?).  

Plotly is great btw, but it runs on javascript. You can use it in Notebooks and for web applications, but I'm not sure whether it would be suited for a desktop app, unless you use a HTML-based frontend. 
>Built on top of the Plotly JavaScript library ([plotly.js](https://plotly.com/javascript/)), plotly enables Python users to create beautiful interactive web-based visualizations that can be displayed in Jupyter notebooks, saved to standalone HTML files, or served as part of pure Python-built web applications using Dash.[^1]

Replacing TeeChart with Scottplot in Wave while sticking with .NET means "only" the charting features need to be rewritten, the backend responsible for reading/writing and analyzing time series does not have to be touched. It's a sizable change, but requires much less effort than writing a whole new app in Python. 

All that said, I'm open to a new Python timeseries app (pyWave?), but not on my own :-) I'd be more than willing to contribute!

[^1]: https://plotly.com/python/getting-started/
